### DISCUSSION TITLE
Debouncing state slices

### DISCUSSION THREAD

**Discussion by francislavoie** [pmndrs/zustand#1179] [Q&A] [answered]
We have some very-high-frequency updated data in the store (big list of data for `react-table` to render, updated live via websockets).

What's the recommended way to wrap the `useStore(state => state.tableData)` to debounce it, so we can throttle the re-render frequency?

I haven't seen any discussion surrounding this with Zustand, so I was wondering if there was any particular recommendations.

For Redux, people seem to use functions like this https://stackoverflow.com/a/64799942 to wrap the selector in a timer. Would that work fine with Zustand too, or are there caveats I should be aware of? Has anyone written somekind of middleware or plugin to do this that I could take a look at? I haven't found any.

**Comment by dai-shi** [comment, marked as answer]
I think you can subscribe to a store and set local state with debounce.

Like https://github.com/pmndrs/zustand#transient-updates-for-often-occurring-state-changes, but `useState` instead of `useRef`.

**Reply by francislavoie** [reply]
Ah yeah, that makes sense! Didn't think to look at subscribers for this. Thanks!

**Reply by cryptokoala4** [reply]
I have this same exact scenario, and subscribe has helped me skip re-rendering my table on large quantity of individual incoming data from websockets in 10-50ms ranges. 

@dai-shi  For the purpose of understanding, could you explain a little more why .subscribe is able to achieve this behaviour? It appears to have some sort of built in throttle with fast changing states.

**Reply by francislavoie** [reply]
Yeah - I wouldn't mind a bit more of a fleshed out explanation in that section of the README.

**Reply by dai-shi** [reply]
It's more about how React behaves right?

Simple counter example:
```js
import { useReducer, useRef, useEffect } from 'react';

const Counter = () => {
  const countRef = useRef(0);
  useEffect(() => {
    const id = setInterval(() => {
      ++countRef.current;
    }, 100);
    return () => clearInterval(id);
  }, []);
  return <>count: {countRef.current}</>;
};

const App = () => {
  const forceUpdate = useReducer((c) => c + 1, 0);
  return (
    <>
      <div><button onClick={forceUpdate}>Re-render</button</div>
      <div><Counter /></div>
    </>
  );
};
```

Usually, this is anti-pattern in React. Zustand is designed to be used with r3f, so "transient updates" is important for apps such as games. In general, it's not very recommended for Reactish apps, especially because concurrent rendering may result in unexpected behaviors.
